### ISSUE TITLE
Adjust resources of csi-node-driver DaemonSet

### ISSUE THREAD

**Issue by PKizzle** [tigera/operator#2412]
## Current Behavior

It is not possible to specify the resources of the csi-node-driver DaemonSet using the Installation reference for the helm chart.

## Possible Solution

Add a `csiNodeDriverDaemonSet` option to the Installation reference that overrides the resource field for the mentioned DaemonSet

## Context

All other resources created by the tigera operator have an option to adjust the resources except the csi-node-driver

**Comment by rene-dekker** [comment]
Looks like someone just did this recently: https://github.com/tigera/operator/pull/2331

**Comment by yoholiao** [comment]
> 

```yml
tigeraOperator:
  image: tigera/operator
  version: v1.29.0
  registry: quay.io
calicoctl:
  image: docker.io/calico/ctl
  tag: v3.25.0
```
i use this version . `csiNodeDriverDaemonSet` config not work。

**Comment by hajdukda** [comment]
Missing an option to override CSI tolerations :( Running calico 3.24

**Comment by PKizzle** [comment]
The first issue I encountered is that the container names are wrong in the updated `installations.operator.tigera.io` CRD. But even after fixing that with the latest tigera-operator v1.29.1 version any overrides are just ignored and never applied.

**Comment by tmjd** [comment]
The PR Rene mentioned (#2331) has not been picked back to v3.24 or v3.25. Both those version have received an update that sets the default for the csi-node-driver tolerations to all like the calico-node daemonset redeives.
Anyone trying to use the `csiNodeDriverDaemonSet` will need to wait for v3.26 at this time.

**Comment by stdmje** [comment]
Any news about when v3.26 will be released? I need to set tolerations on csi-node-driver pods.

**Comment by tmjd** [comment]
Probably in the first half of April.

**Comment by dops-at** [comment]
Any update on when v3.26 will be released?

**Comment by tmjd** [comment]
It looks like currently it is expected the week of the May 15th.

**Comment by Pavaningithub** [comment]
Even after upgrading to v3.26.1(latest as of today), I'm still unable to override the resources of the `csi-node-driver`
Anybody able to override csi-node-driver?


ArgoCD Application which I'm using to configure calico-operator

```yaml
project: default
source:
  repoURL: 'https://docs.tigera.io/calico/charts'
  targetRevision: v3.26.1
  helm:
    values: >-
      installation:
        kubernetesProvider: EKS
        csiNodeDriverDaemonSet:
          spec:
            template:
              spec:
                containers:
                - name: csi-node-driver
                  resources:
                    requests:
                      memory: "64Mi"
                      cpu: "200m"
                    limits:
                      memory: "128Mi"        
  chart: tigera-operator
destination:
  server: 'https://kubernetes.default.svc'
  namespace: calico-operator
syncPolicy:
  automated:
    prune: true
    selfHeal: true
  syncOptions:
    - CreateNamespace=true
    - Replace=true
````

csi-node-driver pod details
```yaml
args:
  - --nodeid=$(KUBE_NODE_NAME)
  - --loglevel=$(LOG_LEVEL)
env:
  - name: LOG_LEVEL
    value: warn
  - name: KUBE_NODE_NAME
    valueFrom:
      fieldRef:
        apiVersion: v1
        fieldPath: spec.nodeName
image: docker.io/calico/csi:v3.26.1
imagePullPolicy: IfNotPresent
name: calico-csi
resources: {}
securityContext:
  allowPrivilegeEscalation: true
  capabilities:
    drop:
      - ALL
  privileged: true
  runAsGroup: 0
  runAsNonRoot: false
  runAsUser: 0
  seccompProfile:
    type: RuntimeDefault
terminationMessagePath: /dev/termination-log
terminationMessagePolicy: File
volumeMounts:
  - mountPath: /var/run
    name: varrun
  - mountPath: /etc/calico
    name: etccalico
  - mountPath: /csi
    name: socket-dir
  - mountPath: /var/lib/kubelet
    mountPropagation: Bidirectional
    name: kubelet-dir
  - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
    name: kube-api-access-qqqsl
    readOnly: true
```

**Comment by tscn** [comment]
@Pavaningithub I managed to configure `nodeAffinity` on `csi-node-driver` DaemonSet. However, it only worked by editing the installation resource directly. I have no clue yet why it does not work through helm for me.

**Comment by quantumdecoherence** [comment]
I also cannot pass any resources to the csi-node-driver. using this config: 

`  - name: tigera-operator
    namespace: tigera-operator
    chart: projectcalico/tigera-operator
    version: "v3.26.1"
    values:
      - installation:
          kubernetesProvider: EKS
<similar resource settings for calico-node, typha, and kube-controller left out. These work fine>
          csiNodeDriverDaemonSet:
            spec:
              template:
                spec:
                  containers:
                    - name: csi-node-driver
                      resources:
                        requests:
                          memory: 9M
                          cpu: 10m
                        limits:
                          memory: 128Mi
                          cpu: 100m
`
describing the installation gives me this Status.Message:
`Invalid Installation provided: Installation spec.CSINodeDriverDaemonSet is not valid: spec.Template.Spec.Containers["csi-node-driver"] is invalid: container "csi-node-driver" is invalid. No containers are expected`

`spec.Template.Spec.Containers.name` is a required value when passing any spec, and it must be set to "csi-node-driver" according to helm.

**Comment by tmjd** [comment]
It looks like the [previous PR](https://github.com/tigera/operator/pull/2331) that added the csiNodeDriverDaemonSet only included tolerations/affinity/nodeselector even though the API seems to include container overrides the code was not added to support resource overrides.
In other words the code currently does not understand any overrides on the specific container(s) for the csi-node-driver daemonset.

I would be happy to review a PR to add resource overrides for the containers for the csi-node-driver daemonset.

**Comment by yeahservice** [comment]
@tmjd I've opened a PR that should fix the issue. Testing it locally works for me. Please take a look.

**Comment by d-afanasiev** [comment]
Any updates on this issue?

**Comment by tmjd** [comment]
Following the chain of related PRs for this it looks like this was fixed and made available in operator v1.33.1.
